stage: add an rhsm.facts stage

supakeen opened the pull request:
This stage allows to leave behind a build ID on the built image to later identify the job used to build that image.

So, this PR is an initial PR to centralize discussion on what we want and/or need to store in the image. The directory name was chosen out of an earlier discussion but is something I might want to think about some more.

In this initial idea the minimum amount of data is stored on the system, only an identifier that could be used by tooling to later look up the job that built the image in the jobqueue. We might want to store more information directly though I'd like to keep this file as *small* as possible both in scope and in size.

After this PR is merged subsequent PRs will be created in `osbuild-composer`, and `image-builder`.

Gundersanne commented:
Nice, so this would be the minimal thing to make it useful longterm imo. But I wonder how composer(/image-builder) aware we wanna make osbuild? Like an arbitrary key-value thing in an easily machine-readable structure is kinda all we need and if that's more palatable for osbuild (maintainers) that could be better?

Otherwise lgtm :)

supakeen commented on stages/org.osbuild.breadcrumb.image-builder:
```diff
@@ -0,0 +1,46 @@
+
+def main(tree, options):
+    path = os.path.join(tree, "usr", "share", "osbuild", "self")
+    file = os.path.join(path, "image-builder.json")
```
I'm undecided on this as well, both on names of the file(s), location of the file(s), and if it should be a single file or separate ones.

Including version information is something I do want to do (though I don't know if all tools can interrogate their own version number).

teg replied:
I feel there are three potential kinds of things to put into the image:

- the manifest (how was the image built) - out of scope for this PR?
- version information about the build tools (in what environment was it built) - also out of scope?
  - the osbuild git sha
  - the host kernel version?
  - the image definition git sha (i.e., composer as long as that's where the image info lives)
- information about what/how the user requested the image - I think this is the most relevant right now
  - is the origin of the image `on-prem`, `marketplace` or `service` (probably not the right labels)?
  - what sort of configuration was applied at a high level? (modified filesystem, non-RHEL content, etc). think about the kind of questions we might want to ask for high-level business analysis (for instance to determine whether certain features are used / useful)

I think a simple JSON file with key/value pairs might give us all we need, but something more clever might be worth it (I suggest focusing on the third point for now though).

gicmo replied:
> the manifest (how was the image built) - out of scope for this PR?

I think that could be a different stage even, and we write the manifest into `self/manifest.json`

> I think a simple JSON file with key/value pairs might give us all we need

Yeah, the question here was more like, how should that file be called and how tight should we pin down the schema for the key-value pairs. I mean we can always have it free form, but I feel that defining a proper schema for it would be better since client tooling will require the key, values to be in an a well defined form and if that exists we might as well put it into the schema here to so clients can rely that it does not change. 

A hybrid approach would be to have some well known information `version` and such well defined in the file and then allow for an `image-builder` section that is free form and we just pass on as-is.

gicmo replied:
See https://github.com/osbuild/osbuild/pull/1060#issuecomment-1190449770

teg commented:
Sounds like what we want is an rhsm facts specific stage. I suggest we make it generic and allow key/value pairs so we can use it for other things than our immediate needs. For instance, I wonder if this would also benefit edge fleet managment.

Re symlink: I would rather just put the file in /etc if we anyway need to symlink it. How would it work for ostree as used by edge though?

gicmo commented:
> Sounds like what we want is an rhsm facts specific stage. 

Yes

> I suggest we make it generic and allow key/value pairs so we can use it for other things than our immediate needs. For instance, I wonder if this would also benefit edge fleet managment.

I have argued against that earlier and I still think it makes sense to pin down the key/values including their types and which ones are required, so that we get some level of type safety and also can rely on certain keys being in a there and in a specific form. 

> Re symlink: I would rather just put the file in /etc if we anyway need to symlink it. How would it work for ostree as used by edge though?

I think `/etc` is in theory the wrong place because this is fixed data describing the image/commit and not something to be editable by the system administrator. Specifically on OSTree systems it would be nice if that was enforce by being in read-only `/usr` (and we avoid potential issues with the etc three-way-merge in case someone does in fact touch the file).

gicmo commented:
> > I suggest we make it generic and allow key/value pairs so we can use it for other things than our immediate needs. For instance, I wonder if this would also benefit edge fleet managment.
> 
> I have argued against that earlier and I still think it makes sense to pin down the key/values including their types and which ones are required, so that we get some level of type safety and also can rely on certain keys being in a there and in a specific form.

In some Wittgensteinian argument with myself: it is a bit funny to encode some specific requirements for a special use case in a more generic `rhsm.facts` stage. Could have one specific to image builder (but that should then maybe even be out of tree and shipped with Composer) or we allow arbitrary k/v pairs but restrict the place where the file can be created so the stage doesn't get "abused" as generic JSON creation stage.

teg commented:
I agree on fixing the location, and I think having it in `/usr/share/osbuild/rhsm.facts` or something like that, with a symlink from `/etc/rhsm/facts/osbuild.facts` would make sense. We discussed this, and the main benefit would be for OSTree based images, where this might be useful (used by Edge Fleet Management, as far as I know).
As to what restrictions to place: @gicmo and I discussed this, and I think we landed on supporting whatever the rhsm facts file format does (whether that's generic k/v or if there are any restrictions).

The actual info to put in here, and when to do it, would be decided by the image definitions in composer, and no further updates should be required to the stage when/if we add information.

dvdhrm commented:
I would prefer the symlink from `/etc` as well. This can be easily created during firstboot, or tmpfiles, and can be dropped when subman can read data from `/usr`. We still want empty `/etc` for pristine images, so lets allow for it.

supakeen commented:
The PR was rebased, the stage has been renamed to `org.osbuild.rhsm.facts` and creates arbitrary key/values in a JSON file at `/usr/share/osbuild/rhsm.facts`. This file is then symlinked into `/etc/rhsm/facts/osbuild.facts`.

gicmo commented:
> The PR was rebased, the stage has been renamed to `org.osbuild.rhsm.facts` and creates arbitrary key/values in a JSON file at `/usr/share/osbuild/rhsm.facts`. This file is then symlinked into `/etc/rhsm/facts/osbuild.facts`.

Great, thanks!! One thing: `/usr/share/osbuild/rhsm.facts`, I think we might want a sub-directory in there that is for all things related to the image itself, especially since we already have `/usr/share/osbuild/schemas`. Having random files next to it seems a bit dirty.

supakeen commented:
Added a `self` directory to put this into in `/usr/share/osbuild` (thought I saw it mentioned somewhere) :)

dvdhrm commented on stages/org.osbuild.rhsm.facts:
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/python3
+"""
+The `org.osbuild.rhsm.facts` stage is allows for storing rhsm-facts format
+artifacts in the built image. Keys and values are supplied in options.
```
This reads really weirdly. How about:

```suggestion
This stage allows storing artifact-properties of the Red Hat Subscription Manager (rhsm-facts) in the built image.
```

dvdhrm commented on stages/org.osbuild.rhsm.facts:
```diff
@@ -0,0 +1,66 @@
+
+
+SCHEMA = r"""
+"definitions": {
+  "key": {
+    "type": "string"
+  },
+  "value": {
+    "type": "string"
+  }
+},
```
You can just inline that, or is there a particular reason to have this as separate definitions?

supakeen replied:
There's no specific reason I was planning on perhaps adding a pattern for the allowed RHSM key structure but that's premature.

dvdhrm commented on stages/org.osbuild.rhsm.facts:
```diff
@@ -0,0 +1,66 @@
+"properties": {
+  "facts": {
+    "type": "array",
+    "description": "Array of key/value facts.",
```
```suggestion
    "description": "Array of Red Hat Subscription Manager facts.",
```

dvdhrm commented on stages/org.osbuild.rhsm.facts:
```diff
@@ -0,0 +1,66 @@
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "key": {
+          "$ref": "#/definitions/key"
+        },
+        "value": {
+          "$ref": "#/definitions/value"
+        }
```
As said above, this would become:

```suggestion
        "key": { "type": "string" }
        "value": { "type": "string" }
```

dvdhrm commented on stages/org.osbuild.rhsm.facts:
```diff
@@ -0,0 +1,66 @@
+
+    os.makedirs(path, exist_ok=True)
+
+    with open(file, "w", encoding="utf-8") as f:
```
How about:

```suggestion
    with open(file, "x", encoding="utf-8") as f:
```

This would make this stage fail if the file exists. Note that the current code would silently drop any previous contents. This would catch manifest errors where the stage is specified multiple times, and thus some items might not end up in the resulting image?

dvdhrm commented on stages/org.osbuild.rhsm.facts:
```diff
@@ -0,0 +1,66 @@
+    os.makedirs(os.path.join(tree, "etc", "rhsm", "facts"), exist_ok=True)
+
+    os.symlink(
+        os.path.join("/usr/share/osbuild/self/rhsm.facts"),
+        os.path.join(tree, "etc", "rhsm", "facts", "osbuild.facts"),
```
You use `join` quite inconsistently, imo. Either be fine with manual concatenation, or join all of them. My preferred style is `os.path.join(tree, "etc/rhsm/facts/osbuild.facts")`, but regardless what you decide, the first argument here should be consistent with the second, shouldn't it?

Also, if you define `path = "usr/share/osbuild/self"`, you can use it here and in `os.makedirs()`.

dvdhrm reviewed (changes requested):
I like it much more now! Some improvement suggestions inline, but feel free to disregard them on a case-by-case basis.

supakeen commented:
Thank you for the comments @dvdhrm. I've applied all of the changes.

dvdhrm commented on stages/org.osbuild.rhsm.facts:
```diff
@@ -0,0 +1,54 @@
+
+SCHEMA = r"""
+"properties": {
+  "facts": {
+    "type": "array",
+    "description": "Array of Red Hat Subscription Manager facts.",
+    "items": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "key": { "type": "string" },
+        "value": { "type": "string" }
+      }
+    }
+  }
```
Valid configuration will be like:

```json
{
    "facts": [
        {
            "key": "foo",
            "value": "content1"
        },
        {
            "key": "bar",
            "value": "content2"
        }
    ]
}
```

The same could be achieved with a more simple configuration like:

```json
{
    "facts": {
        "foo": "content1",
        "bar": "content2"
    }
}
```

The latter is much more dense and would use `additionalProperties = True` to implement. It would not allow duplicate keys (probably a good thing), and would not allow easy extensions of the `facts` key, since it would be fully defined already.

I am fine with the initial approach as well. So just something to think about, in case you prefer the second approach as well.

supakeen replied:
I prefer the second approach as well. I believe rhsm facts cannot be nested instead they have some logic to automatically nest dotted keys.

dvdhrm commented on stages/org.osbuild.rhsm.facts:
```diff
@@ -0,0 +1,54 @@
+    os.makedirs(os.path.join(tree, "etc/rhsm/facts"), exist_ok=True)
+
+    os.symlink(
+        file,
```
This is not correct, is it? `file` is prefixed with `tree`, which should not leek into the symlink, right? My previous comment was a bit confusing, probably. I explicitly dropped the `tree` prefix from my suggestion when sharing the path.

supakeen replied:
It's not correct, I got overzealous.

dvdhrm reviewed (commented):
The symlink looks wrong to me. You now leak the tree-path into the symlink. Note that the destination of symlinks is stored verbatim in the filesystem. It is *not* resolved at creation time.

supakeen commented:
The symlink was wrong after the review changes, adjusted that. I much agree that the `additionalProperties` approach is a lot cleaner in regards to manifest so I've changed it to use that as well.

dvdhrm reviewed (approved):
I like it a lot now! Lets see how the others will tear it apart ;)

teg commented on stages/org.osbuild.rhsm.facts:
```diff
@@ -0,0 +1,47 @@
+
+
+def main(tree, options):
+    path = os.path.join(tree, "usr/share/osbuild/self")
```
The commit message does not mention `self`, is this a left-over? FWIW, I would vote against `self` just because this doesn't *necessarily* describe just the image, it is just RHSM facts, whatever they might be.

dvdhrm replied:
What would you use instead? "self" was discussed beforehand in this PR.

I think it kinda depends on what these facts define. "self" was picked as a subdir in "osbuild" to distinguish data describing the host image from data installed by rpm as part of osbuild to build new images on this host.

This is not so much about what the data describes, but rather where the data comes from. In this case, it states the data was part of the host-image, rather than part of an installed osbuild-rpm.

teg replied:
Ah, that makes sense. I misunderstood and missed the discussion.

teg commented on stages/org.osbuild.rhsm.facts:
```diff
@@ -0,0 +1,47 @@
+    os.makedirs(path, exist_ok=True)
+
+    with open(file, "x", encoding="utf-8") as f:
+        json.dump(options["facts"], f)
```
Is the format really arbitrary JSON? If so: 👍 , but if there are restrictions, we should enforce them.

dvdhrm replied:
"Any user-defined JSON files that override or add system facts to determine entitlement compatibility. Any facts files must end in .facts."

That's all I got.

Gundersanne replied:
Should we reinforce the key prefix somehow, or have that as a separate option? It's not a strict requirement but it's definitely useful to have all our facts share the same prefix, also to avoid any conflicts between other custom facts. Then again we can also just make sure that composer always uses the prefix.

teg replied:
If the tool places no restriction I think we should keep the policy in composer. @gicmo?

gicmo replied:
We could make it easy to use from composer by having a `prefix` option that would get applied to all the keys. But I guess we can also just use the go json annotation a la `Version string 'json:"image_builder.version"'`

Gundersanne reviewed (approved):
I like it :)

teg reviewed (approved):
Thanks for this! Looking great!

gicmo commented on stages/org.osbuild.rhsm.facts:
```diff
@@ -0,0 +1,47 @@
+
+
+def main(tree, options):
+    path = os.path.join(tree, "usr/share/osbuild/self")
```
I think @teg was not a fan `osbuild/self`. Maybe we just put it in  `/usr/share/osbuild/rhsm.facts/`?

dvdhrm replied:
Lets maybe discuss this separately (slack?) There's now 3 threads about that in here, and all have at one point yielded and agreed on "self" ;)

gicmo reviewed (changes requested):
Only the target dir is something to sort of before merging.

gicmo commented:
Let's go with self for now, we can still change it in a follow up.
